### DISCUSSION TITLE
JSON DB Schema

### DISCUSSION THREAD

**Discussion by weizhongzou** [C-Accel-CRIPT/Python-SDK#178] [Q&A]
Not quite sure what “$ref” and “$def” means in line 483.

**Comment by nh916** [comment]
### JSON DB Schema
`$ref` and `$def` are for working with the DB JSON Schema. `$ref` within a JSON block means to refer to the definition of that particular schema, so if we have nested structures the DB JSON Schema code is kept DRY, by just writing it out once and then referring to it wherever needed. I believe the `$def` keyword in JSON Schema is used to define reusable schemas. This can be useful for avoiding duplication of schema definitions, and for making schemas easier to read and maintain.

For the API class we use it to tell the JSON DB Schema to validate our JSON with this particular schema. We use the [JSONSchema package](https://python-jsonschema.readthedocs.io/en/stable/) to validate our JSON with the DB Schema.

### CRIPT JSON DB Schema
There is a naming convention throughout the code of “ProjectPost” that is used for validation when sending an HTTP POST request to create a new project, and “ProjectPatch” that is used for validation when sending an HTTP Patch request to update a project.

Feel free to look into the [official JSON Schema documentation](https://json-schema.org/) to learn more.
